### DISCUSSION TITLE
Getting InsecureSecretKey error on release build

### DISCUSSION THREAD

**Discussion by InstinctBas** [rwf2/Rocket#2528] [Questions] [answered]
This is very likely user error but trying out my project on a test server with the release build, I'm getting an `InsecureSecretKey` error when starting up.

I've generated my secret key as instructed on the help pages with `openssl rand -base64 32`

My `Rocket.toml` file contains:
```
[release]
...
secret_key = "<base64 string pasted in>"
```

This worked on v4 (but with `[production]` as the entry.

Any ideas what I'm doing wrong?

Thanks, Bas

**Comment by InstinctBas** [comment]
One thing I was wondering about is that a `/` character is present in my base64 string, could it be that this needs to be escaped?

**Comment by SergioBenitez** [comment]
If you're using the default config provider, then that suffices. You just need to make sure that the Rocket.toml file is in the CWD or a parent of where you are ~~building~~ running.

**Reply by InstinctBas** [reply]
Building? Is the Rocket.toml now included into the compiled binary? I thought the release version read from the rocket.toml file installed alongside it?

**Reply by SergioBenitez** [reply]
Apologies, yes, I meant to say running. 

To be clear, the Rocket.toml file doesn't need to be in any particular location or even named in a particular way. You can always specify the config file to use with ROCKET_CONFIG. But by default, Rocket will look for the file in the current working directory, which may or may not be the directory where the binary is.

See https://rocket.rs/v0.5-rc/guide/configuration/#default-provider for what Rocket does by default.

**Reply by InstinctBas** [reply]
ok, then I'm not sure why it's not working. As far as I can see the Rocket.toml is in the right place. Is there any way I can verify it was loaded correctly such as adding a print command in my code that prints out the configuration?

else if it is loading correctly, I don't get why it doesn't accept the secret key.

**Reply by SergioBenitez** [reply]
Are you using the default config provider or did you change it? If you changed it, then you have full responsibility for all values, and Rocket doesn't do any of its default configuration. But if you didn't (perhaps you could share how you build your instance of Rocket?) I would try a few things:

Set `ROCKET_LOG_LEVEL=debug`. See if the log messages provide any insight. Verify that the config is what you expect it to be. In particular, check that the secret key is `provided`.

Along with the above, set the path to the file directly using `ROCKET_CONFIG=/path/to/Rocket.toml`.

**Reply by InstinctBas** [reply]
No, as far as I know it's still set to defaults, but maybe there is a step i've missed in the conversion from v4?

Sadly this project is on a private git instance.

I'll try and set the debug level and see what happens, won't be able to do that until later today/this week, we're right in the middle of a migration of our gitlab server :)

**Reply by SergioBenitez** [reply]
Were you able to resolve this?

**Reply by InstinctBas** [reply]
No not yet, the migration has thrown up some issues that I need to resolve first

**Reply by InstinctBas** [reply]
I tried setting both `ROCKET_LOG_LEVEL` and `ROCKET_CONFIG` but it is not giving me much more.. it's like it's not even attemping to log the Rocket.toml. 

Do I need to do something extra to tell it to use the default config provider?

**Reply by InstinctBas** [reply]
Rocket is running in a docker image, so checking the image:
```
[root@localhost 90000]# docker run --rm jbapi-rust printenv
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
HOSTNAME=49c7334f9c04
ROCKET_LOG_LEVEL=normal
ROCKET_CONFIG=/srv/app/Rocket.toml
HOME=/root
```

Output when deploying the image:
```
jbapi-9_0_000_jbapi.1.rduwfp1aj3fx@localhost.localdomain    | Error: InsecureSecretKey(Profile(Uncased { string: "release" }))
jbapi-9_0_000_jbapi.1.rduwfp1aj3fx@localhost.localdomain    | deregistering event source from poller
```

My `Rocket.toml` (this is a test server so don't mind sharing):
```
[default]
address = "0.0.0.0"
port = 8000
workers = 2
limits = { forms = 32768 }

[debug]
log = "normal"

[release]
# log = "critical"
log = "normal"
secret_key = "yTuWLyU9obwIfOsgi9mJd2MWPWirKu24eN5Ba/5Ox2Y="
```

**Reply by InstinctBas** [reply]
Ok, running it locally in release mode it works as is should, so it's something to do with access..

**Reply by InstinctBas** [reply]
also yes, log => log_level

**Reply by InstinctBas** [reply]
@SergioBenitez looking at the source code, the secret key is checked before we print any information about the configuration. That makes sense as at this point the configuration is still being amended (i.e. we may create a secret key if we're not in production).

This does mean that if we exit at an error at this place (i.e. https://github.com/SergioBenitez/Rocket/blob/master/core/lib/src/rocket.rs#L551), we haven't output any useful information.

Seeing the code here:
```
        #[cfg(feature = "secrets")]
        if !config.secret_key.is_provided() {
            if config.profile != Config::DEBUG_PROFILE {
                return Err(Error::new(ErrorKind::InsecureSecretKey(config.profile.clone())));
            }
```
It is clear that the error is caused because no configuration is loaded, and seeing this is only happening in docker I'm guessing at this point I have an rights issue...

The error is misleading btw, no secret key is provided...

**edit** ah, I'm wondering if dockers file replacement is to blame here... that explains why before the error was about incorrect characters... 
I assume that if the secret key is incorrect it will be cleared, hence `is_provided()` returning false not just when no secret key is present, but also if an incorrect secret key is present?

**Reply by SergioBenitez** [reply]
> @SergioBenitez looking at the source code, the secret key is checked before we print any information about the configuration. That makes sense as at this point the configuration is still being amended (i.e. we may create a secret key if we're not in production).
> 
> 
> 
> This does mean that if we exit at an error at this place (i.e. https://github.com/SergioBenitez/Rocket/blob/master/core/lib/src/rocket.rs#L551), we haven't output any useful information.
> 
> 
> 
> Seeing the code here:
> 
> ```
> 
>         #[cfg(feature = "secrets")]
> 
>         if !config.secret_key.is_provided() {
> 
>             if config.profile != Config::DEBUG_PROFILE {
> 
>                 return Err(Error::new(ErrorKind::InsecureSecretKey(config.profile.clone())));
> 
>             }
> 
> ```
> 
> It is clear that the error is caused because no configuration is loaded, and seeing this is only happening in docker I'm guessing at this point I have an rights issue...
> 
> 
> 
> The error is misleading btw, no secret key is provided...

Please let me know what the root cause is once you determine it. If we can provide helpful information during launch to avoid whatever kind of issue it is you're encountering, I'd love to add that in.

**Reply by InstinctBas** [reply, marked as answer]
Ok I figured it out, kind off, and the problem isn't Rocket but it is docker. And it turns out it has been broken for years, it is just that v4 was happy to generate a secret key if none was provided in production (not a complaint, I agree v5 this loophole has been closed).

So we deploy our solution to our servers through docker, the docker image has default configuration files that shouldn't work in production. On the servers we have the proper configuration file for that server and when deploying the image we mount those files.

It was mounting our config file, but it wasn't mounting the rocket one. It just silently failed.

I think the cause is because mounting files isn't really supported well in docker, it's kind of a loop hole thing. Its meant for mounting folders, so I'm guessing because both config files are in the same folder it is just skipping over it.

I've got it to work now by taking a different approach to this.

My only suggestion for Rocket atm is that it should log a bit more state when there is a failure. It's a bit of a catch-22 I guess seeing we're so early on in the setup and Rocket hasn't had time to evaluate a whole lot of state.

I think for this particular case it would be nice if it actually logged the key value with the error. I guess that is a potential security risk if it logs an actually valid key, but it would also have told us more of what is going on.

**Reply by SergioBenitez** [reply]
With https://github.com/SergioBenitez/Rocket/commit/615e70fdad61354e9148977f6650f599a017944c, Rocket, with debug logs enabled, will now print the source location (provenance) of every configuration variable it reads. This would have immediately told you that it was not reading your `Rocket.toml` file. Hopefully this helps folks in the future.
